### PR TITLE
refactor: add custom ULID type for sqlalchemy

### PR DESCRIPTION
**Pull request by Panaetius**
this allows us to have ULID on ORMs. In the DB it will still be a string
/deploy

### CODE REVIEW

**Comment by Panaetius** [comment]
This change is part of the following stack:

- #326
    - #341
        - #342
            - #345
                - #346 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: components/renku_data_services/utils/sqlalchemy.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        """Transform value for storing in the database."""
+        if value is None:
+            return None
+        return str(value)
+
+    def process_result_value(self, value: str | None, dialect: Dialect) -> ULID | None:
+        """Transform string from database into ULID."""
+        if value is None:
+            return None
+        return cast(ULID, ULID.from_str(value))
```
Comment (leafty) [inline components/renku_data_services/utils/sqlalchemy.py:25]: non-blocking: should we explain why we `cast` here?

**Review by leafty — APPROVED** [review]
Nice!

**Review by leafty — APPROVED** [review]

**Review by olevski — APPROVED** [review]

**Review by olevski — DISMISSED** [review]

**Comment by Panaetius** [comment]
This change is part of the following stack:

- #346 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Comment by Panaetius** [comment]
I reworked this a bit and combined this PR with #348

**Review by olevski — DISMISSED** [review]

**Review by olevski — APPROVED** [review]

**Pull request merged**